Update draft-ietf-tls-external-psk-guidance.md (#69)
tlswg/external-psk-design-team: draft-ietf-tls-external-psk-guidance.md

  RFC8446:

informative:
  RFC8773:
  RFC7925:
  RFC6066:
  RFC6614:
  RFC4122:
  RFC2865:
  I-D.ietf-tls-dtls13:
  I-D.ietf-tls-ctls:

>   I-D.ietf-tls-external-psk-importer:

Reviewer: "SHOULD [use importers]" seems like it would push this one into the normative section.

Reviewer 2: Agreed.

This document provides usage guidance for external Pre-Shared Keys (PSKs)

> in Transport Layer Security (TLS) version 1.3 as defined in RFC 8446.

Reviewer 2: ```suggestion
in Transport Layer Security (TLS) 1.3 as defined in RFC 8446.
```

This document lists TLS security properties provided by PSKs under certain
assumptions, and then demonstrates how violations of these assumptions lead
to attacks. This document discusses PSK use cases and provisioning processes.
This document provides advice for applications to help meet these assumptions.
This document also lists the privacy and security properties that are not

> provided by TLS version 1.3 when external PSKs are used.

Reviewer 2: ```suggestion
provided by TLS 1.3 when external PSKs are used.
```

--- middle

# Introduction

This document provides guidance on the use of external Pre-Shared Keys (PSKs)

> in Transport Layer Security (TLS) version 1.3 {{RFC8446}}. This guidance also
> applies to Datagram TLS (DTLS) version 1.3 {{I-D.ietf-tls-dtls13}} and
> Compact TLS version 1.3 {{I-D.ietf-tls-ctls}}; however, for readability only
> TLS is used in the discussion. External PSKs are symmetric secret keys provided
> to the TLS protocol implementation as external inputs; they are provisioned
> out-of-band. This document lists

Reviewer 2: ```suggestion
in Transport Layer Security (TLS) 1.3 {{RFC8446}}. This guidance also
applies to Datagram TLS (DTLS) 1.3 {{I-D.ietf-tls-dtls13}} and
Compact TLS 1.3 {{I-D.ietf-tls-ctls}}. For readability, this document uses
the term TLS to refer to all such versions. External PSKs are symmetric 
secret keys provided to the TLS protocol implementation as external inputs. 
External PSKs are provisioned out-of-band. This document lists
```

# PSK Security Properties {#sec-properties}

When using external PSK authentication, the use of previously established

> keys allows TLS endpoints to authenticate the endpoint identities.  However,

Reviewer 2: ```suggestion
PSKs allows TLS endpoints to authenticate the endpoint identities.  However,
```

these keys do not provide privacy protection of endpoint identities
(see {{endpoint-privacy}}), nor do they provide
non-repudiation (one endpoint in a connection can deny the conversation).

> Protection of endpoint identities and protection against an endpoint denying

Reviewer 2: I would recommend deleting this entire sentence, for two reasons: (1) the server identity is still leaked even if certificates are used (and ECH is not used), and (2) there may be other future forms of authentication that prohibit deniability. It seems reasonable to just say "PSK-based auth has these properties," and nothing more.

Author: I think you are suggesting:

When using external PSK authentication, these TLS 1.3 does not provide privacy protection ...

If that is what you mean, then I a comfortable.
If you mean something else, please provide words.

Reviewer 2: Yeah, that's what I mean.

obvious weaknesses here:

1. Any group member can impersonate any other group member.
1. If PSK is combined with DH, then compromise of a group member that knows
the resulting DH shared secret will enable the attacker to read (and modify) traffic.
1. If PSK is not combined with DH, then compromise of any group member allows the
attacker to passively read (and actively modify) all traffic.

Additionally, a malicious non-member can reroute handshakes between honest group members

> to connect them in unintended ways, as described below. Note that this class of attack

Reviewer 2: I'd revert this text too. @kaduk's comment on selfie doesn't seem to be in the scope of that attack, so IMO the previous text more accurately describes things.

Reviewer: I don't think that even strict use of SNI renders the class of attack "not possible", full stop.
But your point that this is tangential and we may not need to go into quite so much detail seems reasonable.

Author: I do not see how reverting addresses the original comment.

> to connect them in unintended ways, as described below. Note that this class of attack
> can sometimes be detected by each group member including the SNI extension {{RFC6066}}
> and terminating the connection on mismatch between the presented SNI value and the
> receiving member's known identity. The SNI extension names the server identity,
> but the group members that acting as clients are not identified by the SNI
> extension. See {{Selfie}} for details.

Reviewer: ```suggestion
to connect them in unintended ways, as described below. Note that a partial mitigiation
against this class of attack is available: each group member includes the SNI extension {{RFC6066}}
and terminates the connection on mismatch between the presented SNI value and the
receiving member's known identity.  See {{Selfie}} for details.
```

# Security Considerations {#security-con}

Security considerations are provided throughout this document.  It bears

> repeating that there are concerns related to the use of external PSKs related
> to proper identification of the TLS version 1.3 endpoints and additional
> risks when the external PSKs are know to a group.

Reviewer 2: ```suggestion
repeating that there are concerns related to the use of external PSKs regarding
proper identification of TLS 1.3 endpoints and additional risks when external 
PSKs are known to a group.
```

It is NOT RECOMMENDED to share the same PSK between more than one client and server.
However, as discussed in {{use-cases}}, there are application scenarios that may
rely on sharing the same PSK among multiple nodes. {{I-D.ietf-tls-external-psk-importer}}
helps in mitigating rerouting and Selfie style reflection attacks when the PSK
is shared among multiple nodes. This is achieved by correctly using the node
identifiers in the ImportedIdentity.context construct specified in
{{I-D.ietf-tls-external-psk-importer}}. It is RECOMMENDED that each endpoint

> selects one globally unique identifier and uses it in all PSK handshakes. The

Reviewer: Apparently I didn't find this noteworthy the first time around, but reading it again, "RECOMMENDED to always use a single fixed GUID" has pretty lousy privacy properties.  Even if we think this is the right tradeoff based on the functionality needs, maybe we should acknowledge that it has poor privacy properties.

Reviewer 2: Couldn't hurt! Good suggestion.

Author: s/It is RECOMMENDED that each/One solution is that/

Reviewer 2: Yep, that'd work.
